pkg/controller/statusmanager/machineconfig_status.go

```
@@ -0,0 +1,248 @@
+			machineConfigsBeingRemoved[role].Insert(deleted.UnsortedList()...)
+		}
+	}
+	if !reflect.DeepEqual(machineConfigsBeingRemoved, status.machineConfigsBeingRemoved) {
+		status.machineConfigsBeingRemoved = machineConfigsBeingRemoved
+	}
+	var annotateUpdate bool
+	// When there are new rendered machine configs, update the annotation cache.
+	for role, newlyRenderedMachineConfigs := range newlyRenderedMachineConfigMap {
+		if _, ok := renderedMachineConfigMap[role]; !ok {
+			renderedMachineConfigMap[role] = sets.Set[string]{}
+			renderedMachineConfigMap[role].Insert(newlyRenderedMachineConfigs.UnsortedList()...)
+			annotateUpdate = true
+		} else {
+			new := newlyRenderedMachineConfigs.Difference(renderedMachineConfigMap[role])
+			if len(new) == 0 {
+				continue
+			}
+			renderedMachineConfigMap[role].Insert(new.UnsortedList()...)
+			annotateUpdate = true
+		}
+	}
+	if annotateUpdate {
+		status.renderedMachineConfigs = renderedMachineConfigMap
+		if err := status.setLastRenderedMachineConfigState(renderedMachineConfigMap); err != nil {
+			klog.Errorf("failed to set rendered machine config state: %v", err)
+		}
+	}
+	status.Unlock()
+	mcPools := &mcfgv1.MachineConfigPoolList{}
+	err = status.client.ClientFor("").CRClient().List(ctx, mcPools)
+	if err != nil {
+		klog.Errorf("failed to retrieve machine config pools: %v", err)
+		return
+	}
+	status.SetFromMachineConfigPool(mcPools.Items)
+}
+
+func (status *StatusManager) SetFromMachineConfigPool(mcPools []mcfgv1.MachineConfigPool) {
+	status.Lock()
+	defer status.Unlock()
+	// The status.renderedMachineConfigs is a non-nil map at the time when SetFromMachineConfigPool method is invoked.
+	for role, machineConfigs := range status.renderedMachineConfigs {
+		pools, err := status.findMachineConfigPoolsForLabel(mcPools, map[string]string{platform.MachineConfigLabelRoleKey: role})
+		if err != nil {
+			klog.Errorf("failed to get machine config pools for the role %s: %v", role, err)
```

Please return the errors.